spot_driver/spot_driver/calibrated_reregistered_hand_camera_depth_publisher.py

```diff
@@ -0,0 +1,233 @@
+can be populated into the underlying bosdyn API, then the default registered depth image
+could be used without this script.
+
+While it is possible to calibrate the gripper camera entirely under the hood, there is no way to
+only set the intrinsic/extrinsic parameters (to preserve control over calibration parameters)
+
+This script assumes you have a calibration generated with /spot_wrapper/spot_wrapper/calibration
+"""
+import argparse
+from typing import Optional
+
+import bdai_ros2_wrappers.process as ros_process
+import cv2
+import numpy as np
+import open3d as o3d
+import yaml
+from bdai_ros2_wrappers.context import wait_for_shutdown
+from cv_bridge import CvBridge
+from sensor_msgs.msg import Image
+
+
+class CalibratedReRegisteredHandCameraDepthPublisher:
+    def __init__(
+        self,
+        calibration_path: Optional[str] = None,
+        tag: str = "default",
+        robot_name: str = "spot",
+        topic_name: str = "depth_registered/hand_custom_cal/image",
+        undistort: bool = False,
+    ):
+        self.node = ros_process.node()
+
+        if calibration_path is None:
+            self.node.get_logger().warning("No calibration path found, using default calibration")
+            self.calibration = {
+                "camera_matrix_depth": np.zeros(9).tolist(),
+                "dist_coeffs_depth": np.zeros(5).tolist(),
+                "camera_matrix_rgb": np.zeros(9).tolist(),
+                "dist_coeffs_rgb": np.zeros(5).tolist(),
+                "depth_t_rgb_R": np.zeros(9).tolist(),
+                "depth_t_rgb_T": np.zeros(3).tolist(),
+                "rgb_image_dim": np.zeros(2).tolist(),
+            }
+        else:
+            self.calibration = extract_calibration_parameters(calibration_path=calibration_path, tag=tag)
+
+        self.calibration["robot_name"] = robot_name
+        self.calibration["topic_name"] = topic_name
+        self.calibration["undistort"] = undistort
+
+        for param_name, param_val in self.calibration.items():
+            self.node.get_logger().info(f"Setting the parameter {param_name} to a default value of {param_val}")
+            self.node.declare_parameter(param_name, param_val)
+            self.calibration[param_name] = self.node.get_parameter(param_name).value
+            if param_name in ["camera_matrix_depth", "camera_matrix_rgb", "depth_t_rgb_R"]:
+                self.calibration[param_name] = np.array(self.calibration[param_name]).reshape((3, 3))
+            elif "dim" not in param_name and "name" not in param_name and "undistort" not in param_name:
+                self.calibration[param_name] = np.array(self.calibration[param_name])
+
+        self.cv_bridge = CvBridge()
+
+        raw_depth_topic = f"/{self.calibration['robot_name']}/depth/hand/image"
+        self.node.get_logger().info(f"Creating subscriber to raw depth at {raw_depth_topic}")
+        self.raw_depth_img_sub = self.node.create_subscription(
```

Comment:
I would create the subscriber last to prevent a possible race condition where you receive an image before you create the  publisher

Reply:
re-ordered in latest commit